_posts/2021-03-14-Julia-1.6-what-has-changed-since-1.0.md

```diff
@@ -0,0 +1,737 @@
+```julia
+# Method for if x is given as components
+foo(x1, x2; a=10, b=20, c=30, op=+) = foo(Bar(op(x1, x2)); a, b, c)
+```
+
+This does come with the requirement to separate keyword arguments from positional arguments by `;`, but I have always done this, and the [BlueStyle guide requires it](https://github.com/invenia/BlueStyle#keyword-arguments).
+It feels like we are fully leveraging the distinction of keyword from positional arguments by allowing this.
+Where as the previous motivation for this distinction (vs e.g. C# and Python that allow any positional argument to be passed by name), was to make the name of positional arguments not part of the public API, thus avoiding changing it being a breaking change.
+This new feature feels more ... featureful.
+
+The same syntax can be used to create `NamedTuple`s.
+```julia
+julia> product = ["fries", "burger", "drink"];
+
+julia> price = [2, 4, 1];
+
+julia> (;product, price)
+(product = ["fries", "burger", "drink"], price = [2, 4, 1])
+```
+This is particularly cool for constructing `NamedTuple`s of `Vector`s, which is a valid [Tables.jl Table](https://tables.juliadata.org/stable/#Tables.columntable).
+
+It is interesting to note that for the logging macros introduced in Julia 1.0, this is how they have always worked.
+E.g. `@info "message" foo bar` and `@info "message" foo=foo bar=bar` display the same, which has always felt natural.
+
+### Lowering of `'`
+In Julia 1.0 `'` was lowered directly into a call to `Base.adjoint`.
+This meant it was impossible to redefine what `'` meant in your current module.
+Now it lowers to a call to `var"'"`, which is something you can overload.
+In `Base` it is overloaded to `Base.adjoint` so nothing has changed.
+People often think it is cool to overload this in automatic differentiation packages so that `f'(x)` gives you the derivative of `f(x)`.
+
+Furthermore, `'` can now have Unicode suffixes added to it, to define a new suffix operator.
+For example, the one in `Base` which makes `A'ᵀ` give `transpose(A)` (rather than the adjoint/hermitian transpose which is returned by `A'`).
+
+## Performance: References to the Heap from the Stack
+This was promised in 2016 as a feature for 1.0 (released 2018), but we actually didn't get it until 1.5 (released 2020) with [#33886](https://github.com/JuliaLang/julia/issues/33886).
+In short, the process of allocating memory from the heap is fairly slow\*, whereas allocating memory on the stack is basically a no-op.
+Indeed, Julia benchmarking tools don't count allocations on the stack as allocations at all.
+One can find extensive write ups of heap vs. stack allocations and how it works in general (though some mix the C specific factors with the CPU details).
+In Julia all mutable objects live are allocated on the heap.
```

Comment:
This needs an asterisk, occasionally the compiler is able to not create a mutable struct if it can prove that you won't notice.